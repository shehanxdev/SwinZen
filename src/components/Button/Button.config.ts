export const animatedButtonBoundries = {
  initialWidth: '100%',
  initialBorderRadius: 10,
  loadingWidth: '14%', //Magic number which closes value to get a rounded view
  loadingBorderRadius: 100,
};
