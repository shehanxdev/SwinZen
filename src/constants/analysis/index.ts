export * from './analysis.enum';
export * from './pro-tips-content';
