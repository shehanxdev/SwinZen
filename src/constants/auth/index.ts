export * from './otp-data.enum';
