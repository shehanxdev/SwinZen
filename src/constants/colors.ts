/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable @typescript-eslint/no-redeclare */

//NOTE:: Since an enum member cannot have a numeric name prefixed with "Sz"
enum Primary {
  Sz900 = `#070707`,
  Sz800 = `#0C1E0F`,
  Sz700 = `#16501D`,
  Sz600 = `#28642A`,
  Sz500 = `#45734A`,
  Sz400 = `#69AA67`,
  Sz300 = `#8DBE8B`,
  Sz200 = `#C0DBBF`,
  Sz100 = `#F0F7F0`, //NOTE::figma contain two different colors(#F0F7F0, #C0DBBF) with the border. #C0DBBF neeeds to be removed since it does not consider within the colors
}

enum Neutral {
  Sz1000 = `#283E36`, //NOTE::not available within the design system colors
  Sz900 = `#212529`,
  Sz800 = `#343A40`,
  Sz700 = `#495057`,
  Sz600 = `#6C757D`,
  Sz500 = `#ADB5BD`,
  Sz400 = `#CED4DA`,
  Sz300 = `#DEE2E6`,
  Sz200 = `#E9ECEF`,
  Sz100 = `#F8F9FA`, //NOTE::figma contain two different colors(#F8F9FA, #DEE2E6) with the border. #DEE2E6 neeeds to be removed since it does not consider within the colors
  Black = `#000000`,
  White = `#FFFFFF`,
}

enum Secondary {
  Sz900 = `#F65815`,
  Sz800 = `#F87944`,
  Sz700 = `#F99469`,
  Sz600 = `#FAA17C`,
  Sz500 = `#FBBCA1`,
  Sz400 = `#FDD7C7`,
  Sz300 = `#FDE4D9`,
  Sz200 = `#FFF2ED`,
  Sz100 = `#FFE3B0`, //NOTE::figma contain two different colors(##FFE3B0, #FFF9F6) with the border. #FFF9F6 neeeds to be removed since it does not consider within the colors
}

/* Currently this is available within the design system as a Secondary value. Since two different colors cannot exist in there, this has to be added to the design system */
enum Tertiary {
  Sz900 = `#A2FD2F`,
  Sz800 = `#B5FD59`,
  Sz700 = `#C4FE7A`,
  Sz600 = `#CBFE8A`,
  Sz500 = `#DAFEAC`,
  Sz400 = `#E9FFCD`,
  Sz300 = `#F0FEDE`,
  Sz200 = `#F8FFEF`,
  Sz100 = `#DAFEAC`, //NOTE::figma contain two different colors(#DAFEAC, #FBFFF7) with the border. #FBFFF7 neeeds to be removed since it does not consider within the colors
}

enum Transparency {
  Sz100 = `#212529`,
  Sz80 = `#212529cc`,
  Sz64 = `#212529a3`,
  Sz56 = `#2125298f`,
  Sz40 = `#21252966`,
  Sz24 = `#2125293d`,
  Sz16 = `#21252929`,
  Sz8 = `#21252914`,
  Sz4 = `#2125290a`,
  full = `#ffffff00`, //NOTE:: This is full transpapernt color. Not available under the design system.
}

enum Error {
  SzMain = `#E35F5F`, //Not defined within the design system but being used within the screens :(
}

const SwingZenThemePalette = {
  Primary,
  Neutral,
  Secondary,
  Tertiary,
  Transparency,
  Error,
};

export type Color = Primary | Neutral | Secondary | Tertiary | Transparency | Error;

export const Color = {
  ...SwingZenThemePalette,
};
