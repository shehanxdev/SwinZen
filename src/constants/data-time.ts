export const dateFormats = {
  default: 'YYYY-MM-DD',
  calenderHeader: 'MMMM YYYY',
  rangeDatePicker: 'MMM DD',
  scratchedDateAndTime: 'HH:MM, DD/MM/YY',
  singleDatePicker: 'ddd, MMM DD',
  textInput: 'DD / MM / YY',
  extrasMarketHeader: 'dd/MM/YY',
  finalResultsHeader: 'dd/MM/YYYY',
  dayAndDate: 'ddd, DD',
  timestamp: 'DD/MM/YYYY HH:mm:ssPM Z',
  advertisedStartTime: 'ddd, DD MMM YYYY HH:mm PM',
  saveFileTimestamp: 'DD_MM_YYYY_HH_mm_ss',
};
