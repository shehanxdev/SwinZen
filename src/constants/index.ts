export * from './auth';
export * from './error';
export * from './colors';
export * from './data-time';
export * from './navigation-routes';
export * from './regular-expressions';
export * from './runtime-environment';
export * from './static-info';
export * from './typography';
