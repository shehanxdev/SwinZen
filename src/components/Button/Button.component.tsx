import React, { useEffect } from 'react';
import { useMemo, useState } from 'react';
import { View } from 'react-native';
import { Button as RNPaperButton } from 'react-native-paper';
import Animated, { useAnimatedStyle, useSharedValue, withSpring, withTiming } from 'react-native-reanimated';

import { tw } from '@sz/config';
import { Color, TextVariant } from '@sz/constants';

import { LoadingIndicator } from '../LoadingIndicator';
import { Text } from '../Typography';
import { animatedButtonBoundries } from './Button.config';
import { ButtonProps } from './Button.types';

export function Button({
  backgroundColor = Color.Tertiary.Sz900,
  activeStateBackgroundColor = Color.Tertiary.Sz1000,
  onPress,
  onLongPress,
  textColor = Color.Neutral.Sz1000,
  testID,
  title,
  borderColor = Color.Transparency.full,
  disabledBackgroundColor = Color.Neutral.Sz900,
  disabledTextColor = Color.Neutral.Sz700,
  disabledBorderColor = Color.Neutral.Sz700,
  disabled = false,
  uppercase = true,
  loading = false,
}: ButtonProps) {
  const [isButtonClicked, setIsButtonClicked] = useState<boolean>(false);

  const width = useSharedValue(animatedButtonBoundries.initialWidth); //Has to go with percentages due to a technical blocker
  const borderRadius = useSharedValue(animatedButtonBoundries.initialBorderRadius); //px

  const buttonAnimatedStyles = useAnimatedStyle(() => {
    return { width: width.value, borderRadius: borderRadius.value };
  });

  //handle loading animation when the loading state changes
  useEffect(() => {
    if (loading) {
      width.value = withSpring(animatedButtonBoundries.loadingWidth);
      borderRadius.value = withSpring(animatedButtonBoundries.loadingBorderRadius);
    } else {
      width.value = withTiming(animatedButtonBoundries.initialWidth);
      borderRadius.value = withSpring(animatedButtonBoundries.initialBorderRadius);
    }
  }, [loading]);

  const buttonBackgroundColor = useMemo(() => {
    const buttonBackgroundColor =
      disabled && disabledBackgroundColor
        ? disabledBackgroundColor
        : isButtonClicked
        ? activeStateBackgroundColor
        : backgroundColor;

    return buttonBackgroundColor;
  }, [disabled, disabledBackgroundColor, isButtonClicked]);

  const buttonBorderColor = useMemo(() => {
    const buttonBorderColor = disabled ? disabledBorderColor : borderColor;

    return buttonBorderColor;
  }, [disabled, disabledBorderColor, isButtonClicked]);

  return (
    <Animated.View
      style={[
        tw`flex-row bg-[${buttonBackgroundColor}] rounded-2.5 justify-center shadow-none border border-[${buttonBorderColor}]`,
        buttonAnimatedStyles,
      ]}>
      {loading && (
        <View style={tw`absolute self-center z-1`}>
          <LoadingIndicator size="small" color={textColor} />
        </View>
      )}
      <RNPaperButton
        compact={true}
        uppercase={uppercase}
        disabled={disabled}
        onPress={onPress}
        style={tw`flex-1`}
        contentStyle={tw`h-12`}
        testID={testID}
        onTouchStart={() => {
          setIsButtonClicked(true);
        }}
        onTouchEnd={() => {
          setIsButtonClicked(false);
        }}
        onLongPress={onLongPress}>
        {!loading && (
          <Text variant={TextVariant.Body1SemiBold} color={disabled ? disabledTextColor : textColor}>
            {title}
          </Text>
        )}
      </RNPaperButton>
    </Animated.View>
  );
}
