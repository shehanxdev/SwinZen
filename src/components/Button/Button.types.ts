import React from 'react';
import { Button } from 'react-native-paper';

import { Color } from '@sz/constants';

type RNPaperButtonProps = React.ComponentProps<typeof Button>;

export type WithRNPaperButtonProps = Pick<
  RNPaperButtonProps,
  'onPress' | 'testID' | 'disabled' | 'onLongPress' | 'uppercase'
>;

export interface ButtonProps extends WithRNPaperButtonProps {
  backgroundColor?: Color;
  activeStateBackgroundColor?: Color;
  title: string;
  textColor?: Color;
  disabledBackgroundColor?: Color;
  disabledTextColor?: Color;
  borderColor?: Color;
  disabledBorderColor?: Color;
  loading?: boolean;
}
