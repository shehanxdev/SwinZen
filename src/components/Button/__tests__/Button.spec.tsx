import { fireEvent, render } from '@testing-library/react-native';
import * as React from 'react';

import { Color } from '@sz/constants';

import { Button } from '../Button.component';
import { ButtonProps } from '../Button.types';

describe('Button Component', () => {
  const dummyText = 'dummy text';
  const otherPropSet = [
    {
      backgroundColor: Color.Primary.Sz300,
      activeStateBackgroundColor: Color.Secondary.Sz100,
      textColor: Color.Transparency.Sz24,
    },
    {
      backgroundColor: Color.Neutral.White,
      activeStateBackgroundColor: Color.Neutral.Sz100,
      textColor: Color.Neutral.Sz300,
      disabledBackgroundColor: Color.Neutral.Sz100,
      disabledTextColor: Color.Neutral.Sz100,
      borderColor: Color.Neutral.Sz100,
      disabledBorderColor: Color.Neutral.Sz100,
      disable: true,
    },
    {
      backgroundColor: Color.Tertiary.Sz700,
      activeStateBackgroundColor: Color.Secondary.Sz100,
      fullWidth: true,
    },
  ];
  const testID = 'ButtonTestID';
  const mockOnPressFunction = jest.fn();
  const mockOnLongPressFunction = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  const getRenderedButtonComponent = (props?: Partial<ButtonProps>) =>
    render(<Button {...props} onPress={mockOnPressFunction} testID={testID} title={dummyText} />);

  describe('should render correctly', () => {
    for (const otherProp of otherPropSet) {
      it(`should render correctly with ${otherProp}`, () => {
        const rendered = getRenderedButtonComponent({ ...otherProp });
        const renderedTree = rendered.toJSON();
        expect(renderedTree).toMatchSnapshot();
      });
    }
  });

  it('should find the button via testID', () => {
    const { getByTestId } = getRenderedButtonComponent();
    const foundButton = getByTestId(testID);

    expect(foundButton).toBeTruthy();
  });

  it(`should show relevant text inside the button`, () => {
    const { getByText } = getRenderedButtonComponent();
    const foundTextElement = getByText(dummyText);

    expect(foundTextElement.props.children).toEqual(dummyText);
  });

  it('should not call on press function, if the button IS IN disable state', () => {
    const { getByTestId } = getRenderedButtonComponent({ disabled: true });

    fireEvent.press(getByTestId(testID));
    expect(mockOnPressFunction).not.toHaveBeenCalled();
  });

  it('should call on press function, if the button IS NOT in disable state', () => {
    const { getByTestId } = getRenderedButtonComponent({ disabled: false });

    fireEvent.press(getByTestId(testID));
    expect(mockOnPressFunction).toHaveBeenCalled();
  });

  it('should not call on long press function, if the button IS IN disable state', () => {
    const { getByTestId } = getRenderedButtonComponent({ disabled: true, onLongPress: mockOnLongPressFunction });

    fireEvent.press(getByTestId(testID));
    expect(mockOnPressFunction).not.toHaveBeenCalled();
  });

  it('should call on long press function, if the button IS NOT in disable state', () => {
    const { getByTestId } = getRenderedButtonComponent({ disabled: false, onLongPress: mockOnLongPressFunction });

    fireEvent.press(getByTestId(testID));
    expect(mockOnPressFunction).toHaveBeenCalled();
  });
});
