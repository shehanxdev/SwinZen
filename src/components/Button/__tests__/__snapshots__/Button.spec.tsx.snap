// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render correctly should render correctly with [object Object] 1`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <View
    collapsable={false}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "backgroundColor": "#8DBE8B",
        "borderColor": "#ffffff00",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "elevation": 0,
        "justifyContent": "center",
        "minWidth": "auto",
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 10,
          },
        ]
      }
      testID="ButtonTestID"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            {
              "height": 48,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              {
                "textAlign": "left",
              },
              [
                {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                {
                  "marginHorizontal": 8,
                },
                false,
                {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#2125293d",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
              ]
            }
          >
            dummy text
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Button Component should render correctly should render correctly with [object Object] 2`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <View
    collapsable={false}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#F8F9FA",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "elevation": 0,
        "justifyContent": "center",
        "minWidth": "auto",
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 10,
          },
        ]
      }
      testID="ButtonTestID"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            {
              "height": 48,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              {
                "textAlign": "left",
              },
              [
                {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                {
                  "marginHorizontal": 8,
                },
                false,
                {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#DEE2E6",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
              ]
            }
          >
            dummy text
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Button Component should render correctly should render correctly with [object Object] 3`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
>
  <View
    collapsable={false}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      {
        "backgroundColor": "#C4FE7A",
        "borderColor": "#ffffff00",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 1,
        "elevation": 0,
        "justifyContent": "center",
        "minWidth": "auto",
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 10,
          },
        ]
      }
      testID="ButtonTestID"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            {
              "height": 48,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              {
                "textAlign": "left",
              },
              [
                {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                {
                  "marginHorizontal": 8,
                },
                false,
                {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#283E36",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
              ]
            }
          >
            dummy text
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
